docs/rfcs/0004-cirrus-experiments.md

```diff
@@ -0,0 +1,260 @@
+
+```
+{
+  "featuresOverwrite":  {
+    "Feature1":  true
+    "Feature2":  false
+    "FeatureN":  true
+  }
+}
+```
+
+This endpoint will be changed to only accept POST requests. Requests will be expected to contain the following JSON body, which is compliant with the body of the request that will be sent to Cirrus:
+
+```
+{
+  "client_id":  "4a1d71ab-29a2-4c5f-9e1d-9d9df2e6e449",
+  "context":  {
+    "key1":  "value1",
+    "key2":  {
+      "key2.1":  "value2",
+      "key2.2":  "value3"
+    }
+  }
+}
+```
+
+> **Note** Throughout this document, client_id will be referred to as "experimenter id" so as not to confuse it with the Glean client_id.
+
+The Cirrus reply will be added to the body of the response object, like so:
+```
+{
+ "featuresOverwrite":  {
+    "Feature1":  true
+    "Feature2":  false
+    "FeatureN":  true
+  },
+  "experimentalFeatures":  {
+    "Feature1":  {
+      "Variable1.1":  "valueA",
+      "Variable1.2":  "valueB"
+    },
+    "Feature2":  {
+      "Variable2.1":  "valueC",
+      "Variable2.2":  "valueD"
+    },
+    "FeatureN":  {
+      "VariableN.1":  "valueX",
+      "VariableN.2":  "valueY"
+    }
+  }
+}
+```
+
+> **Note** This request does not need to be authenticated. It only needs to be made using HTTPS.
+
+If "featuresOverride" conflicts with "experimentalFeatures", "experimentalFeatures" will always have higher priority.
+
+### Client side implementation
+
+Client side will be responsible for querying the server for active features and applying treatments based on the obtained data.
+
+#### 1\. Generate an experimenter id
+
+In order for Nimbus to run its targeting logic, it requires an experimenter id to be attached to each request.
+
+The experimenter id must be generated per user. It's important that this is user based to ensure users are assigned to the same features in all of their logged in devices.
+
+That poses an issue for non-logged in users. In these cases, a random UUID will be generated for the experimenter id. (See also question 2. on the "Open questions" section of this document).
+
+Nimbus relies on Glean data for experiment analysis. Therefore the generated experimenter id needs to be included in all Glean pings. The Glean team is working on a new API that will facilitate setting this id, see[  Bug 1848201](https://bugzilla.mozilla.org/show_bug.cgi?id=1848201). The Mozilla VPN application will rely on this API to provide the experiment id to Glean.
+
+Timing: The experimenter id must be set as soon as possible on the application lifecycle, since it is required for all Cirrus requests and should be included in all Glean pings.
+
+#### 2\. Query Nimbus for active features
+
+The /featurelist endpoint is already queried by the application  on initialization and once every hour. Other than these triggers, a feature list request will be made whenever the experiment id is set.
```

Comment:
Good callouts. For the first question, yes that will be a non-fatal error and whatever are the currently active features will remain. I will update the document with this caveat.

For the second question, that is a great point and I believe this will be feature dependant. If the feature is something like "change the design of the vpn toggle button" or something very disruptive like that we probably don't want to apply it while the user is interacting with the app... But if it's something like extra servers in the servers list we may not care as much. We are able to attach metadata to features, so this could be some boolean flag that features have that state whether it can be applied dinamically or not. I will add this caveat as well. 

Finally for the last question, I will double check with Santiago, but I do not think we should notify users for a restart to apply experiments. If features require a restart they will be applied on next restart and that is all. The feature list is persisted, if this is not documented in the RFC I will include that as well.

Reply:
Coming back here after a long time: 1. I talked ot Santiago and he agrees we do not need to notify users for a restart. 2. I think the features themselves will change in the way they are implemented and therefore be applied or not on restart. I think that is not something that neeeds to come as a flag from Cirrus.